### DISCUSSION TITLE
`image` tag - add|remove event listeners (load|error)

### DISCUSSION THREAD

**Discussion by GeorgiyZhuravlev** [clientIO/joint#2349] [Q&A] [answered]
### Introduction

I've implemented a custom shape with the `image` tag in its `markup`.
```
const prototypeProperties = {
  markup: [{
    tagName: 'circle',
    selector: 'body'
  }, {
    tagName: 'image',
    selector: 'user'
  },...],
```
When a certain attr of a shape is set, `xlink:href` of that `image` is updated accordingly, in the setter function.
Works great!

The problem I'm trying to solve is that sometimes images are not accessible (broken link eg).
I'm looking for a way to grab somehow the node of that image to attach `load|error` event listeners, and if error has happened - replace broken `xlink:href` with the hardcoded fallback (default image).

I do have `node` as a third param of a setter function (value, refBBox, node), but don't feel right to add and remove listeners there (though it works great), because setters are triggered quite often, where `image` is obviously created once per shape.

So, I'm curious how to get that node only once, after it's created.
(and remove listeners later, if it's possible to get kind of "on:destroy" event for a shape)

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment, marked as answer]
I would not delegate this logic to an attribute setter.

a) I would try to avoid setting the wrong URL in advance. There are many ways to check if an image is valid (e.g.  [here](https://www.zhenghao.io/posts/verify-image-url)).

b) You can capture the `error` on the paper element and then find the cell view the error originated from (the event listener has to be attached to `capture` phase - explained [here](https://stackoverflow.com/questions/14983988/is-bubbling-available-for-image-load-events)).

```js
paper.el.addEventListener('error', (evt) => {
    const view = paper.findView(evt.target);
    console.log(view.model.get('type'));
}, { capture: true });
```

**Reply by GeorgiyZhuravlev** [reply]
Intercepting errors on paper's "level" - great idea!
Works great! Thank you!
